### DISCUSSION TITLE
🔵 We need your upvotes to move forward! 🎀  ⬆️ 🆙 ⬆️

### DISCUSSION THREAD

**Discussion by yoavbls** [yoavbls/pretty-ts-errors#43] [Announcements]
We need your votes on things we can't do ourselves because it depends on others
(e.g. VSCode and TypeScript teams on Microsoft, JetBrains, etc)
  
**Please help us help everybody by voting in the links below** 💙
  
  
# 1. Make types copyable
> Solves https://github.com/yoavbls/pretty-ts-errors/issues/29 

I made an issue + pull request for VSCode.
VSCode team may reject it because they already rejected requests for new style props before and promised to solve this in a more robust method, which didn't happen in the last 3 years.
So please:
### [⬆️ Upvote or comment here](https://github.com/microsoft/vscode/issues/180496) 

# 2. Allow diagnostics messages to have markdown
>Solves https://github.com/yoavbls/pretty-ts-errors/issues/15, https://github.com/yoavbls/pretty-ts-errors/issues/19, https://github.com/yoavbls/pretty-ts-errors/issues/3 and make us closer to solve: https://github.com/yoavbls/pretty-ts-errors/issues/11, https://github.com/yoavbls/pretty-ts-errors/issues/21, https://github.com/yoavbls/pretty-ts-errors/issues/26, https://github.com/yoavbls/pretty-ts-errors/issues/25

If VSCode allows it we could be able to move the formatting to a language server plugin which will allow us to:
1. Control the order of the original error and the pretty one
2. Hide the original error without losing it in the problems pane in VSCode or for other extensions like [Error Lens](https://marketplace.visualstudio.com/items?itemName=usernamehw.errorlens)
3. Support more platforms more easily like Neovim, Visual Studio and maybe even JetBrains IDEs
### [⬆️ Upvote or comment here](https://github.com/microsoft/vscode/issues/54272)

<a href="#user-content-jetbrains-support" id="jetbrains-support">#</a>
# 3. JetBrains support 
> solves https://github.com/yoavbls/pretty-ts-errors/issues/11.

We saw a huge demand for support WebStorm and other JetBrains IDEs.
This ticket on YouTrack probably got JetBrains attention:
### ~~[⬆️ Upvote or comment here](https://youtrack.jetbrains.com/issue/WEB-60634/Add-support-pretty-ts-errors)~~ ✅ Done  
If you're a person from the relevant JetBrains team, please reach me out 💙  

**edit:** They reached out and eventually chose to develop their own version because WebStorm isn't fully supporting LSP. It'll be released as a built-in feature in the next version! (2023.2) which is fantastic news!
  
**another edit**: It released!
Thank you @tomblachut @anstarovoyt @kricact
<a target="_blank" href="https://www.jetbrains.com/webstorm/whatsnew/">
<img src="https://www.jetbrains.com/webstorm/whatsnew/img/2023.2/Formatted-ts-error-624-min.png" height="300" />
</a>

**Comment by RoyRao2333** [comment]
According to JetBrains IDEs, I don't think they would adopt this plugin since they're [planning to do this on their own](https://youtrack.jetbrains.com/issue/WEB-40945/Pretty-typescript-type-errors#focus=Comments-27-7304740.0-0).

But maybe, you guys can work this out together. 😊

**Comment by lookfirst** [comment]
I'm the person who came up with the idea and filed the [jetbrains ticket](https://youtrack.jetbrains.com/issue/WEB-40945/Pretty-typescript-type-errors) 4+ years ago. Nice to see someone else pick this up and move jetbrains along. Great work. This is a game changer for typescript that'll have a large impact on developer productivity.

**Comment by tomblachut** [comment]
@yoavbls thanks for the mention, though I was just a messenger. @aleksei-berezkin implemented the feature in JetBrains IDEs :)

**Comment by joaopedrodcf** [comment]
Seems there is a PR now to fix point 2 

https://github.com/microsoft/vscode/pull/214051

**Comment by JS778** [comment]
> We need your votes on things we can't do ourselves because it depends on others
> 
> (e.g. VSCode and TypeScript teams on Microsoft, JetBrains, etc)
> 
>   
> 
> **Please help us help everybody by voting in the links below** 💙
> 
>   
> 
>   
> 
> # 1. Make types copyable
> 
> > Solves https://github.com/yoavbls/pretty-ts-errors/issues/29 
> 
> 
> 
> I made an issue + pull request for VSCode.
> 
> VSCode team may reject it because they already rejected requests for new style props before and promised to solve this in a more robust method, which didn't happen in the last 3 years.
> 
> So please:
> 
> ### [⬆️ Upvote or comment here](https://github.com/microsoft/vscode/issues/180496) 
> 
> 
> 
> # 2. Allow diagnostics messages to have markdown
> 
> >Solves https://github.com/yoavbls/pretty-ts-errors/issues/15, https://github.com/yoavbls/pretty-ts-errors/issues/19, https://github.com/yoavbls/pretty-ts-errors/issues/3 and make us closer to solve: https://github.com/yoavbls/pretty-ts-errors/issues/11, https://github.com/yoavbls/pretty-ts-errors/issues/21, https://github.com/yoavbls/pretty-ts-errors/issues/26, https://github.com/yoavbls/pretty-ts-errors/issues/25
> 
> 
> 
> If VSCode allows it we could be able to move the formatting to a language server plugin which will allow us to:
> 
> 1. Control the order of the original error and the pretty one
> 
> 2. Hide the original error without losing it in the problems pane in VSCode or for other extensions like [Error Lens](https://marketplace.visualstudio.com/items?itemName=usernamehw.errorlens)
> 
> 3. Support more platforms more easily like Neovim, Visual Studio and maybe even JetBrains IDEs
> 
> ### [⬆️ Upvote or comment here](https://github.com/microsoft/vscode/issues/54272)
> 
> 
> 
> <a href="#user-content-jetbrains-support" id="jetbrains-support">#</a>
> 
> # 3. JetBrains support 
> 
> > solves https://github.com/yoavbls/pretty-ts-errors/issues/11.
> 
> 
> 
> We saw a huge demand for support WebStorm and other JetBrains IDEs.
> 
> This ticket on YouTrack probably got JetBrains attention:
> 
> ### ~~[⬆️ Upvote or comment here](https://youtrack.jetbrains.com/issue/WEB-60634/Add-support-pretty-ts-errors)~~ ✅ Done  
> 
> If you're a person from the relevant JetBrains team, please reach me out 💙  
> 
> 
> 
> **edit:** They reached out and eventually chose to develop their own version because WebStorm isn't fully supporting LSP. It'll be released as a built-in feature in the next version! (2023.2) which is fantastic news!
> 
>   
> 
> **another edit**: It released!
> 
> Thank you @tomblachut @anstarovoyt @kricact
> 
> <a target="_blank" href="https://www.jetbrains.com/webstorm/whatsnew/">
> 
> <img src="https://www.jetbrains.com/webstorm/whatsnew/img/2023.2/Formatted-ts-error-624-min.png" height="300" />
> 
> </a>
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> ### THANK YOU ALL! 💁🏼‍♀️ ###
> 
> 
> 
>

**Comment by natsuki-engr** [comment]
LSP 3.18 is planned a change that `Diagnostic` allows to be set `MarkupContent` as message.
https://microsoft.github.io/language-server-protocol/specifications/lsp/3.18/specification/#diagnostic

**Reply by kevinramharak** [reply]
I did see that a while back, that would solve the duplicate messages part. We still need the change to make the type blocks copyable, but this would be a big improvement.

**Reply by natsuki-engr** [reply]
If diagnostics includeing `MarkupContent` (`MarkdownString`) is allowed in vscode, how does pretty-ts-errors follows this?
If we can hide the original error from `HoverProvider`, pretty-ts-errors should handle it, but since it's a Diagnostic fix, I suspect we'll need to modify TS's LSP. Is that right?

**Reply by kevinramharak** [reply]
See https://github.com/yoavbls/pretty-ts-errors/issues/145#issue-3510748948.

VS Code has a language service plugin option, which allows us to modify the communication between vs code and typescript. So instead of the current approach, we would intercept the diagnostic messages and change them from strings to `MarkupContent`

**Reply by anstarovoyt** [reply]
TypeScript service doesn't implement LSP protorol, it could work only for TS@Go but it also requires support for that on the server side.

**Reply by kevinramharak** [reply]
> TypeScript service doesn't implement LSP protorol, it could work only for TS@Go but it also requires support for that on the server side.

I'm not sure what you mean, maybe that the `tsserver` does not fully implement the specification, which I think might be correct. But VS Code has a [built in extension](https://github.com/microsoft/vscode/tree/main/extensions/typescript-language-features) that uses `tsserver` to implement all the programming language features needed to support javascript and typescript.

Also notice that there the LSP term is overloaded:
- Language Service Plugin (this is a plugin that augments the TypeScript language **server**)
- Language Service Protocol (the specification itself)
- Language Service Protocol Server (the servers that implement it)
- Language Service Protocol Client (the clients that implement it)

These are all related concepts, but what we are talking about is:
- The **Protocol** will support `MarkdownContent`
- The TypeScript Language **Server** has the option to embed **Plugins**
- We aim to implement a **Plugin** to change the **Server** diagnostics from text to `MarkupContent`

**Reply by anstarovoyt** [reply]
I meant that tsserver itself doesn't use LSP at all. It uses own LSP-like (predecessor) protocol which is uncompatible with LSP. So, to add the support for  `MarkdownContent` you need not only change the commands on the tsserver side (which could be done with the plugin, if the server flexible enough, it has to be checked), but you also need to modify the typescript/javascript vscode extension to actually support the new commands, it wouldn't work automatically because it isn't LSP. 

But, the new version of TypeScript (Go) actually uses LSP and it should be much easier to do that for that version.

**Reply by kevinramharak** [reply]
> I meant that tsserver itself doesn't use LSP at all. It uses own LSP-like (predecessor) protocol which is uncompatible with LSP. So, to add the support for `MarkdownContent` you need not only change the commands on the tsserver side (which could be done with the plugin, if the server flexible enough, it has to be checked), but you also need to modify the typescript/javascript vscode extension to actually support the new commands, it wouldn't work automatically because it isn't LSP.
> 
> But, the new version of TypeScript (Go) actually uses LSP and it should be much easier to do that for that version.

I checked it out, and I understand what you mean. It is indeed not possible to just have a plugin modify the diagnostics, I assumed the vscode typescript language features would massage the responses into an LSP format, which would make it compatible with VSCode, but it doesn't do that at all. It has a completely different approach of implementing the language features, which is very surprising to me.

I guess we have to wait for TS 6, before we can implement this.

**Reply by kevinramharak** [reply]
Seems like its up for preview at https://github.com/microsoft/TypeScript/issues/39459#issuecomment-4524226998 with the https://marketplace.visualstudio.com/items?itemName=TypeScriptTeam.native-preview extension. Have not tried it yet, but if anyone is interested to try it out its here.
